codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt

```diff
@@ -552,24 +561,35 @@ private class ServerHttpBoundProtocolTraitImplGenerator(
     }
 
     private fun serverRenderResponseCodeBinding(
-        binding: HttpBindingDescriptor
+        binding: HttpBindingDescriptor?,
+        defaultCode: Int?
     ): Writable {
-        check(binding.location == HttpLocation.RESPONSE_CODE)
+        check(binding?.location == HttpLocation.RESPONSE_CODE || defaultCode != null)
+
         return writable {
-            val memberName = symbolProvider.toMemberName(binding.member)
-            // TODO(https://github.com/awslabs/smithy-rs/issues/1229): This code is problematic for two reasons:
-            // 1. We're not falling back to the `http` trait if no `output.$memberName` is `None`.
-            // 2. It only works when `output.$memberName` is of type `Option<i32>`.
+            binding?.let {
+                val memberName = symbolProvider.toMemberName(binding.member)
+                rust("let status = output.$memberName")
+                if (binding.member.isOptional) {
```

Comment:
Always go through the symbol provider to determine if you're working with an `Option` or not i.e. use `symbolProvider.toSymbol(binding.member).isOptional()` here.

This is because the fact that a shape is optional in the model does not mean that it will get rendered as an `Option` in Rust. An example: https://github.com/awslabs/smithy/issues/1179